src/dice/add_dice/constant_folding.rb

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+
+class AddDice
+  # 定数畳み込みをする
+  class ConstantFolding
+    def initialize(lhs)
+      @sequence = [lhs]
+    end
+
+    # 演算子と右辺を追加する
+    # @param op [Symbol]
+    # @param round_type [Symbol, nil]
+    # @param rhs [Object]
+    def push(op, round_type, rhs)
+      @sequence.push(op, round_type, rhs)
+    end
+
+    # 定数畳み込みをしたノードを生成する
+    def construct_node
+      fold()
+
+      node = @sequence.shift
+
+      while !@sequence.empty?
```

Comment:
```suggestion
      until @sequence.empty?
```

`while !cond` を `until cond` に書き換えると、否定演算子がなくなって読みやすくなります。

参考：https://techracho.bpsinc.jp/hachi8833/2017_01_10/32333#2-27